File: Core/src/main/java/net/winternetwork/bedwars/core/game/stage/WaitingPlayerStage.java
package net.winternetwork.bedwars.core.game.stage;

import net.winternetwork.bedwars.api.game.flag.Flag;
import net.winternetwork.bedwars.api.game.stage.Stage;
import net.winternetwork.bedwars.api.score.ReplaceableList;
import net.winternetwork.bedwars.core.game.GameSettings;
import net.winternetwork.bedwars.core.game.flag.Flags;

import java.util.Arrays;
import java.util.List;

public class WaitingPlayerStage extends Stage {

    public WaitingPlayerStage() {
        super(
                "Esperando Jogadores.",
                20
        );
    }

    @Override
    public List<Flag> getFlags() {
        return Arrays.asList(
                Flags.NO_PVP,
                Flags.NO_BUILD
        );
    }

    @Override
    public ReplaceableList getScoreboard() {
        return new ReplaceableList(
                Arrays.asList(
                        "&1",
                        "&fMapa: &a<mapa>",
                        "&2",
                        "&fAguardando...",
                        "&fJogadores: &a<online>/<max>",
                        "&3",
                        "&ewww.baianowars.kt"
                ),
                (s, player) -> s
                        .replaceAll("<waiting>", String.valueOf(GameSettings.PLAYERS_TO_START - getOnlinePlayers()))
                        .replaceAll("<mapa>", "Void")
                        .replaceAll("<online>", getOnlinePlayers() + "")
                        .replaceAll("<max>", "15")
        );
    }

    @Override
    public void onStageJoin() {
        final int online = getOnlinePlayers();

        broadcast(
                String.format("§eTemos %d jogador%s online.", online, online > 1 ? "es" : ""),
                String.format(
                        "§ePara o jogo começar, precisamos de mais %d.",
                        GameSettings.PLAYERS_TO_START - online
                )
        );
    }

    @Override
    public void onSecondPassed() {
    }

    @Override
    public void onTimeLeft() {
        final int online = getOnlinePlayers();

        if (online >= GameSettings.PLAYERS_TO_START) {
            StageManager.getInstance().callNextStage();
        } else {
            setTimeLeft(getTime());
            onStageJoin();
        }
    }

    @Override
    public void onStageExit() {
        broadcast(
                "§eTemos jogadores suficientes.",
                String.format("§eIniciando o jogo em %d segundos.", getTime())
        );
    }
}

File: Game/src/main/kotlin/net/winternetwork/bedwars/modules/stage/object/WaitingPlayerStage.kt
package net.winternetwork.bedwars.modules.stage.`object`

import net.winternetwork.bedwars.api.game.stage.Stage
import net.winternetwork.bedwars.api.score.ReplaceableList
import net.winternetwork.bedwars.game.settings.GameSettings
import net.winternetwork.bedwars.modules.stage.StageModule
import net.winternetwork.bedwars.modules.stage.flag.Flags

class WaitingPlayerStage(stageModule: StageModule) : Stage(
    "Aguardando jogadores",
    20
) {
    private val stageManager = stageModule.stageManager

    override val flags = listOf(
        Flags.NO_PVP_FLAG,
        Flags.NO_BREAK_FLAG,
        Flags.NO_BUILD_FLAG,
        Flags.NO_GENERATOR_FLAG
    )

    override val scoreboard = ReplaceableList(
        listOf(
            "&1",
            "&fMapa: &a<mapa>",
            "&2",
            "&fAguardando...",
            "&fJogadores: &a<online>/<max>",
            "&3",
            "&ewww.baianowars.kt"
        )

    ) {
        mapOf(
            "<waiting>" to "${GameSettings.PLAYERS_TO_START - onlinePlayers}",
            "<mapa>" to "Gadolandia",
            "<online>" to "$onlinePlayers",
            "<max>" to "15"
        )
    }

    override fun onStageJoin() {
        broadcast(
            "§eTemos %d jogador%s online.".format(
                onlinePlayers, if (onlinePlayers > 1) "es" else ""
            ),
            "§ePara o jogo começar, precisamos de mais %d.".format(
                GameSettings.PLAYERS_TO_START - onlinePlayers
            )
        )
    }

    override fun onTimeLeft() {
        if (onlinePlayers >= GameSettings.PLAYERS_TO_START)
            stageManager!!.next()
        else {
            timeLeft = time
            onStageJoin()
        }
    }

}